sound/soc/sof/topology.c

```diff
@@ -3629,6 +3632,9 @@ static void sof_complete(struct snd_soc_component *scomp)
 	 * IPC. It may be overwritten by alsa-mixer after booting up
 	 */
 	snd_sof_cache_kcontrol_val(scomp);
+
+	/* Destroy all pipeline components. They will be set up dynamically when a PCM is started */
+	sof_pipelines_destroy_all(scomp);
```

Comment:
wait, I am lost here.

Is this first creating the pipelines, then destroying them and then we re-allocate/set them up again?

What's the gain then?

Reply:
@plbossart at the tme of topology loading we set up all the widgets and then destory it at the end of topology loading but only at first boot. This will be helpful in determining if there are any errors in topology.

During subsequent boots after runtime/system PM, only active pipelines will be set up.

Reply:
@ranj063 you should really explain this point, as well as clarify the benefits of the 'on-demand' approach. I am personally not getting what the extra complexity gets us, unless the firmware has secret sauce to turn things off.

We've also had feedback in the past that the firmware should maintain its own context and do the save/restore itself. So wondering overall if we are going in the right direction.

Reply:
@plbossart I think I've improved comments and commit messages a bit now. Please do take a look.